alchemist/alchemist-swingui/src/main/java/it/unibo/alchemist/boundary/gui/isolines/conrec/Conrec.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright (C) 2010-2019, Danilo Pianini and contributors
+ * listed in the main project's alchemist/build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.boundary.gui.isolines.conrec;
+
+import java.util.Objects;
+
+/**
+ * Conrec a straightforward method of contouring some surface represented a regular
+ * triangular mesh.
+ *
+ * Ported from the C++ code by Nicholas Yue (see above copyright notice).
+ * See <a href="http://paulbourke.net/papers/conrec">http://paulbourke.net/papers/conrec</a> for full description
+ * of code and original C++ source.
+ *
+ * author  Bradley White
```

Comment:
Ugh

Super nope.

What's the license of this code? We can't use random code from the Internet, is this GPL-compatible?

Plus, copy-pasting code inside a project is **very** bad. Is there any library with this functionality?

Reply:
I looked for libraries with this particular algorithm, but didn't find anything. There are other libraries but with different algorithms not well explained. 
 
That code comes with a "copyright notice" on the top of the file (which seems to have been replaced by our copyright when I copy pasted the file on the project), in which authors grant permission to use the code as long as that notice is retained. Plus, they grant permission to modify the code "provided that such modifications are not distributed without the explicit consent of the authors". I actually have modified the code but for the purpose of making it respect our checkstyle and pmd rules. Plus, I don't think I am distributing those modifications, am I? No other license is released with the source code. The original comment:
```java
/*
 * Conrec.java
 *
 * Created on 5 August 2001, 15:03
 *
 * Copyright (c) 1996-1997 Nicholas Yue
 *
 * This software is copyrighted by Nicholas Yue. This code is base on the work of
 * Paul D. Bourke CONREC.F routine
 *
 * The authors hereby grant permission to use, copy, and distribute this
 * software and its documentation for any purpose, provided that existing
 * copyright notices are retained in all copies and that this notice is included
 * verbatim in any distributions. Additionally, the authors grant permission to
 * modify this software and its documentation for any purpose, provided that
 * such modifications are not distributed without the explicit consent of the
 * authors and that existing copyright notices are retained in all copies. Some
 * of the algorithms implemented by this software are patented, observe all
 * applicable patent law.
 *
 * IN NO EVENT SHALL THE AUTHORS OR DISTRIBUTORS BE LIABLE TO ANY PARTY FOR
 * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT
 * OF THE USE OF THIS SOFTWARE, ITS DOCUMENTATION, OR ANY DERIVATIVES THEREOF,
 * EVEN IF THE AUTHORS HAVE BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * THE AUTHORS AND DISTRIBUTORS SPECIFICALLY DISCLAIM ANY WARRANTIES, INCLUDING,
 * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE IS PROVIDED ON AN
 * "AS IS" BASIS, AND THE AUTHORS AND DISTRIBUTORS HAVE NO OBLIGATION TO PROVIDE
 * MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
 */
```

Reply:
Yes you are. You are putting it on a public website, violating the license. This code must be isolated, incapsulated in a library, and then used.